### PR TITLE
Studio: Use appdata file to store last push and pull timestamp

### PR DESCRIPTION
**Pull request by katinthehatsite**
## Related issues

Closes https://github.com/Automattic/dotcom-forge/issues/9987
Closes https://github.com/Automattic/dotcom-forge/issues/10027

## Proposed Changes

This PR stores the data related to the last pull and push times in the `appdata-v1.json` file instead of using local storage. It also ensures that the timestamps are updated when the user uses the pull and push operation multiple times.

## Testing Instructions

* Start Studio app `STUDIO_SITE_SYNC=true npm start`
* On Studio site, navigate to the `Sync` tab
* Connect a WPCOM site (bonus point to also connect a staging site along the way)
* Hover over the `Pull` and `Push` buttons and confirm that you can see `You have not pulled / pushed this site yet.` tooltip over each button
* Open `/Users/{yourUser/Library/Application Support/Studio/appdata-v1.json` and confirm that the connected site has `lastPullTimestamp` and `lastPushTimestamp` are set to `null` on a newly connected site
* Start either a push or pull operation
* Once it is finished, observe that the tooltip changed and now displays how long ago you pushed or pulled a site
* Open `/Users/{yourUser/Library/Application Support/Studio/appdata-v1.json` and confirm that the connected site has `lastPullTimestamp` or `lastPushTimestamp` (depending on which operation you triggered) set to a specific timestamp
* Try repeating a pull or push operation with triggering an error and confirm that the timestamp does NOT get reset if the pull or push is not successful
* Finally, repeat the same operation and confirm that tooltip time gets updated

## Pre-merge Checklist

- [ ] Have you checked for TypeScript, React or other console errors?

### CODE REVIEW

File: src/hooks/sync-sites/use-sync-push.ts
Comment (katinthehatsite) [inline src/hooks/sync-sites/use-sync-push.ts]: I am doing these returns to be able to update tooltips on the last push and pull only when the operation is successful and keep them with the same values if the operation fails. This seemed to make sense to me but happy to change those

**Comment by katinthehatsite** [comment]
Sidenote: I am still fixing unit tests that are failing but feel free to give it a first pass of review. Ignore the tests though, I will fix it in a minute

Updated: everything fixed up now!

File: src/hooks/use-pull-push-timestamps.ts
Comment (katinthehatsite) [inline src/hooks/use-pull-push-timestamps.ts]: The tooltip text does not accept promises so I am storing it with this state. If there is a better approach, happy to consider it. I tried using this logic directly within `sync-connected-sites` but I think it was cluttering that file too much

File: src/hooks/use-pull-push-timestamps.ts
Comment (sejas) [inline src/hooks/use-pull-push-timestamps.ts]: Not a strong opinion, but why saving it as a String?
If it were a database, I would save it as a number. Like this `new Date().getTime()` instead of `new Date().toISOString()`

File: src/hooks/use-pull-push-timestamps.ts
Comment (sejas) [inline src/hooks/use-pull-push-timestamps.ts]: I think we need to pass the current locale, to translate the time difference to the user's language.
You can grab it from `const { locale } = useI18nData();`

File: src/components/sync-connected-sites.tsx
Comment (sejas) [inline src/components/sync-connected-sites.tsx]: Why not executing this code inside the "pushSite" ?
The same comment for pullSite.

File: src/hooks/sync-sites/use-sync-push.ts
Comment (sejas) [reply src/hooks/sync-sites/use-sync-push.ts]: Instead of returning true and false, we can call `await refreshTooltip( selectedSite.id, connectedSite.id, 'push' );` in the success case inside this hook.
Or maybe these hooks can accept a onSuccess callback.

File: src/hooks/use-pull-push-timestamps.ts
Comment (sejas) [reply src/hooks/use-pull-push-timestamps.ts]: in an ideal world, we could access `connectedSite.lastPullTimestamp` and `connectedSite.lastPushTimestamp` without the needing to fetch the information from the file.

the `getLastSyncTimeWithType` callback shouldn't need to access the siteId and connectedSiteId, and it would receive the whole connectedSite or directly the timestamp. In fact, it  could live directly in SyncSitesProvider.

I didn't test it, but ideally we use that information directly and every time we need to update the timestamp, we save it to the fiel and also update the apps state.

**Review by sejas — COMMENTED** [review]
@katinthehatsite , it's looking good, and feature wise it's working as described.

My main suggestion is getting rid of `fetchConnectedSite` to display the value of `getLastSyncTimeWithType`. I don't think is necessary, but it's true that I didn't test the refactor myself. Let me know what you think.

File: src/hooks/use-pull-push-timestamps.ts
Comment (katinthehatsite) [reply src/hooks/use-pull-push-timestamps.ts]: Yeah I initially had that in https://github.com/Automattic/studio/pull/714/commits/177dd5df11597cdf905e3cd86371b47b2fba212a but doing some research, I saw that `to.ISOString` was a common way to save dates that would format it as `YYYY-MM-DDZHH:mm:ss` so I went with that approach. I think it makes it look more straightforward that it is a date. It is not a strong preference on my end though 👍

File: src/hooks/use-pull-push-timestamps.ts
Comment (katinthehatsite) [reply src/hooks/use-pull-push-timestamps.ts]: I see, I think this issue is already present outside of this PR. I tried passing just the `locale` but it did not work well for translating dates. I added a small hook and used `formatDistance` function that we already had. This is what it looks like now:

Screenshot 2024-12-04 at 10 32 38 AM

To test this, I added the following line:

```
            "You pulled this site %s ago.": [
                "Pobrałeś tę witrynę %s temu."
            ]
```
to the `studio-pl.jed.json` and loaded the app in Polish. Let me know what you think!

Changes applied in: https://github.com/Automattic/studio/pull/714/commits/f04bc1366748704bbe6ae67803ee4c73c290299c

File: src/components/sync-connected-sites.tsx
Comment (katinthehatsite) [reply src/components/sync-connected-sites.tsx]: Done in: https://github.com/Automattic/studio/pull/714/commits/ebf3a4bec7a112cdba42c89a9a80b18922c009e3

Initially, I did it this way to keep the timestamp/tooltip related logic in one place but after the refactor, it is not necessary anymore.

**Comment by katinthehatsite** [comment]
@sejas It is ready for another round! I applied the suggestions to make the code more straightforward. Let me know what you think 👍

File: src/hooks/use-pull-push-timestamps.ts
Comment (katinthehatsite) [reply src/hooks/use-pull-push-timestamps.ts]: Did a bunch of refactor in https://github.com/Automattic/studio/pull/714/commits/ebf3a4bec7a112cdba42c89a9a80b18922c009e3 to simplify data flow 👍

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (fredrikekelund) [inline src/hooks/sync-sites/sync-sites-context.tsx]: I struggled with all the interdependencies between the different "sub-hooks" here when I worked on https://github.com/Automattic/studio/pull/716, so I can see the appeal in not making this function depend on `connectedSites`.

However, if we can find a way to not introduce the new `updateSingleConnectedWpcomSite` IPC handler, then I think it would be a big win. IMO, with this solution, we are adding complexity to `ipc-handlers.ts` to resolve the issue that is complex state inter-dependencies in this file.

File: src/hooks/sync-sites/use-sync-pull.ts
Comment (fredrikekelund) [inline src/hooks/sync-sites/use-sync-pull.ts]: Instead of passing `connectedSite` all the way in here, could we store it (or just the ID) in the pull state object, perhaps?

EDIT: We already have access to `remoteSiteId`. Maybe that's enough?

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (fredrikekelund) [inline src/hooks/sync-sites/sync-sites-context.tsx]:
```suggestion
			onPullSuccess: ( siteId ) =>
				updateSiteTimestamp(
					connectedSites.find( ( { id } ) => siteId === id ),
					'pull'
				),
```
This doesn't work out-of-the-box, because TS complains that `Array.prototype.find` could return undefined. Still, it simplifies the solution as a whole, because we don't need to pass `connectedSite` around inside `use-sync-pull.ts`.

We could solve the TS error by making `updateSiteTimestamp` return early if `site` is undefined.

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (fredrikekelund) [inline src/hooks/sync-sites/sync-sites-context.tsx]: Same comment here as for `onPullSuccess`.

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (fredrikekelund) [inline src/hooks/sync-sites/sync-sites-context.tsx]: Just curious: was there a specific reason for storing ISO dates instead of timestamps?

File: src/hooks/use-format-localized-timestamps.tsx
Comment (fredrikekelund) [inline src/hooks/use-format-localized-timestamps.tsx:3]: This function seems to be used in conjunction with `date-fns`'s `formatDuration` function when looking at `src/hooks/use-expiration-date.ts`.

Should we do something similar here? I also see that date-fns has a built-in `formatDistance` function ([docs](https://date-fns.org/v4.1.0/docs/formatDistance)). I'm not familiar with that one, though—maybe there's a good reason for us not to use it.

**Review by fredrikekelund — COMMENTED** [review]
Works as expected 👍 However, I recognize some issues in the code (in `SyncSitesProvider` specifically) from when I worked on https://github.com/Automattic/studio/pull/716. If we can find a way to resolve those, I think it'd be great.

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (fredrikekelund) [reply src/hooks/sync-sites/sync-sites-context.tsx]: Here's an idea that could potentially work

```ts

	const [ connectedSiteToUpdate, setConnectedSiteToUpdate ] =
		useState< [ SyncSite[ 'id' ], 'push' | 'pull' ] >();

	const { pullSite, isAnySitePulling, isSiteIdPulling, clearPullState, getPullState } = useSyncPull(
		{
			pullStates,
			setPullStates,
			onPullSuccess: ( siteId ) => setConnectedSiteToUpdate( [ siteId, 'pull' ] ),
		}
	);

	const [ pushStates, setPushStates ] = useState< SyncSitesContextType[ 'pushStates' ] >( {} );
	const { pushSite, isAnySitePushing, isSiteIdPushing, clearPushState, getPushState } = useSyncPush(
		{
			pushStates,
			setPushStates,
			onPushSuccess: ( siteId ) => setConnectedSiteToUpdate( [ siteId, 'push' ] ),
		}
	);

	useEffect( () => {
		if ( ! connectedSiteToUpdate ) {
			return;
		}

		const updatedConnectedSites = connectedSites.map( ( s ) => {
			if ( s.id === connectedSiteToUpdate[ 0 ] ) {
				return {
					...s,
					[ connectedSiteToUpdate[ 1 ] === 'pull' ? 'lastPullTimestamp' : 'lastPushTimestamp' ]:
						new Date().toISOString(),
				};
			}

			return s;
		} );

		setConnectedSites( updatedConnectedSites );
		getIpcApi().updateConnectedWpcomSites( updatedConnectedSites );
	}, [ connectedSiteToUpdate ] );
```

File: src/hooks/use-format-localized-timestamps.tsx
Comment (sejas) [inline src/hooks/use-format-localized-timestamps.tsx]: Let's remove this `_locale` variable and all its dependencies like `useI18nData`.

```suggestion
function formatTimeDistance( duration: Duration ): string {
```

File: src/hooks/use-format-localized-timestamps.tsx
Comment (sejas) [reply src/hooks/use-format-localized-timestamps.tsx:3]: I think we use the `format` in `use-expiration-date.ts` to display a specific format that contains `days and hours`, and `hours and minutes` together in the same sentence. Not sure if we need a specific format in this case, or if it's ok using the default format.

I tested it in Polish and it seems to be working well.

Screenshot 2024-12-05 at 11 18 26

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (sejas) [inline src/hooks/sync-sites/sync-sites-context.tsx]: For the case where the timestamp is undefined we were displaying "1 minute ago".
I think this could be solved by checking if timestamp is a nullish value.

```suggestion
		( timestamp: string | null | undefined, type: 'pull' | 'push' ): string => {
			if ( ! timestamp ) {
```

**Review by sejas — APPROVED** [review]
@katinthehatsite , overall it works great. I found two small edge cases.

1. Now this is so efficient that it doesn't re-render when the tooltip is visible, displaying the always the first time, and not the current time. So if the user is in the tab during 5 minutes it will be displaying 1 minute all the time.

A possible solution is creating a DynamicTooltip in a separate file like this:

```ts
import Tooltip, { TooltipProps } from './tooltip';

function DynamicTooltip( {
	getTooltipText,
	children,
	...props
}: PropsWithChildren< TooltipProps & { getTooltipText: () => string } > ) {
	const [ tooltipText, setTooltipText ] = useState( '' );

	const handleMouseEnter = () => {
		const text = getTooltipText();
		setTooltipText( text );
	};

	return (
		<Tooltip text={ tooltipText } placement="top-start" { ...props }>
			<div onMouseEnter={ handleMouseEnter }>{ children }</div>
		</Tooltip>
	);
}
```
then you can use it like this:

```ts
<DynamicTooltip
	getTooltipText={ () =>
		getLastSyncTimeText( connectedSite.lastPullTimestamp, 'pull' )
	}
	disabled={ isOffline }
>
....
</DynamicTooltip>
```

2. When the `connectedSites.lastPullTimestamp` or the equivalent push is undefined, then it displays the "1 minute" time instead of `You pulled this site %s ago.`. 
This is possibly a corner case that will experience only developers and early testers. But let's fix it anyway. I left a possible solution.

Approving it as it works as expected after fixing the two edge cases I mentioned before and the code looks cleaner. Although there are a couple of suggestions from Fredrik we may want to look into before merging.

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (katinthehatsite) [reply src/hooks/sync-sites/sync-sites-context.tsx]: I explained a bit here: https://github.com/Automattic/studio/pull/714#discussion_r1868916607

From my research, it seemed to be a common practice but it is not a strong preference on my end.

File: src/hooks/use-format-localized-timestamps.tsx
Comment (katinthehatsite) [reply src/hooks/use-format-localized-timestamps.tsx]: I did some cleanup here: https://github.com/Automattic/studio/pull/714/commits/201425f7820cd8c02caf2d2b31720df1623dd208

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (katinthehatsite) [reply src/hooks/sync-sites/sync-sites-context.tsx]: Nice catch - fixed in https://github.com/Automattic/studio/pull/714/commits/870df3712c6318609f6f80df01883e68be27bcf7

**Comment by katinthehatsite** [comment]
Thanks for the review @sejas ! 

> Now this is so efficient that it doesn't re-render when the tooltip is visible, displaying the always the first time, and not the current time. So if the user is in the tab during 5 minutes it will be displaying 1 minute all the time.

Nice, I like that solution. Alternative, I think would be some sort of refresher function that I had in https://github.com/Automattic/studio/pull/714/commits/7b0bcd701134c5ad318b9165796697fe81bc5b56 but that would require completely different approach. I think this approach is simple and elegant. I implemented it in https://github.com/Automattic/studio/pull/714/commits/b3f54408371d9fb21db8834847fe58f525a6c09f

> When the connectedSites.lastPullTimestamp or the equivalent push is undefined, then it displays the "1 minute" time instead of You pulled this site %s ago..

Nice catch, fixed the types for this 👍

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (katinthehatsite) [reply src/hooks/sync-sites/sync-sites-context.tsx]: It is something that I am testing now, it might work but we would need to ensure that we also pass selectedSiteId as a site with the same remoteSiteId can be attached to a couple of different local sites. However, I do like the idea of not passing connectedSite. I will experiment with this a bit more.

Comment (fredrikekelund) [reply src/hooks/sync-sites/sync-sites-context.tsx]: I know I left a few different comments on this topic, but I would start with [this one](https://github.com/Automattic/studio/pull/714#discussion_r1871021613), since I think the concept is pretty clear.

> ensure that we also pass selectedSiteId

Good point. Assuming we get that "for free" with the current solution (because of how the closures are structured), I think it would be an additional benefit if we could make that dependency clearer.

Comment (katinthehatsite) [reply src/hooks/sync-sites/sync-sites-context.tsx]: Implemented it in https://github.com/Automattic/studio/pull/714/commits/40724600c3658b59c8a39e6b72aba8777c271246 since we have localSiteId available there already. Thanks for the suggestion 👍

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (katinthehatsite) [reply src/hooks/sync-sites/sync-sites-context.tsx]: Implemented in https://github.com/Automattic/studio/pull/714/commits/40724600c3658b59c8a39e6b72aba8777c271246

File: src/hooks/sync-sites/use-sync-pull.ts
Comment (katinthehatsite) [reply src/hooks/sync-sites/use-sync-pull.ts]: Responded to this feedback in https://github.com/Automattic/studio/pull/714/commits/40724600c3658b59c8a39e6b72aba8777c271246

File: src/hooks/sync-sites/sync-sites-context.tsx
Comment (fredrikekelund) [reply src/hooks/sync-sites/sync-sites-context.tsx]: I don't want to overcomplicate the review of this PR, but with the changes in https://github.com/Automattic/studio/commit/40724600c3658b59c8a39e6b72aba8777c271246, we are even closer to implementing a solution like this ☝️

The benefit being that we can remove `updateSingleConnectedWpcomSite` from `ipc-handlers` if we update the timestamp in a `useEffect` that reacts to a state change instead of in a memoized callback (like we do now).

Comment (katinthehatsite) [reply src/hooks/sync-sites/sync-sites-context.tsx]: Thanks for the suggestion, I like that we can potentially re-use what is already in there! I tested this approach briefly and I think it is something that could be done but there are some details that would need to be taken care of.  I also think it is okay to have that handler in the `ipc-handlers` for now. It might become handy when we might need to update a specific connected site for some other reason (we also have a similar function for[ updating a local site](https://github.com/Automattic/studio/blob/a5392f7bffafddcbc3c1a94b4ae9f53cd96ff40a/src/ipc-handlers.ts#L206).

Let's create a follow-up issue where we can consider this refactor in detail. I suggest that we proceed with merging as is for now considering the timing (we are about to launch the sync feature) and the urgency (we still have lots of issues to get through) and handle this as a follow-up issue. I will open a new issue for this.

Comment (katinthehatsite) [reply src/hooks/sync-sites/sync-sites-context.tsx]: Created follow up here: https://github.com/Automattic/dotcom-forge/issues/10085

Comment (katinthehatsite) [reply src/hooks/sync-sites/sync-sites-context.tsx]: > I don't want to overcomplicate the review of this PR, but with the changes in https://github.com/Automattic/studio/commit/40724600c3658b59c8a39e6b72aba8777c271246, we are even closer to implementing a solution like this ☝️

Oh, I just saw the comment but I already merged the changes. I will handle it in the follow up issue that I opened 👍

**Pull request merged**
